solidity/random-beacon/README.adoc

```diff
@@ -1 +1,299 @@
+    ///                   Beacon.
+    /// @param blockNumber Block number at which the relay entry was submitted
+    ///                    to the chain.
+    function __beaconCallback(uint256 relayEntry, uint256 blockNumber) external;
+}
+```
+
+=== Timeouts
+
+There are two timeouts for a relay entry to be provided by a group: soft timeout
+and hard timeout.
+
+The soft timeout is the group size multiplied by the number of blocks for a
+member to become eligible to submit relay entry. Eligibility is not enforced
+on-chain but off-chain clients are expected to agree and follow it.
+
+If no entry was provided within the soft timeout, all operators in the group
+start bleeding and losing their stake. The bleeding increases linearly from 0 to
+the governable slashing amount per operator over time, until the hard timeout is
+reached or until relay entry is submitted by the group.
+
+The time for a single group member to become eligible to submit result and the
+hard relay entry timeout are governable parameters. This gives a chance to start
+with more forgiving penalties and increase them over time. In general, the
+slashing penalty should be proportional to rewards and the frequency of relay
+requests and associated risk.
+
+When the hard timeout is reached, anyone can notify about this fact by calling
+`RandomBeacon.reportRelayEntryTimeout()` function and receiving a notifier
+reward. The group which failed to submit a relay entry is terminated, group
+members are slashed, and if there are still active groups in the beacon, another
+group is selected and tasked with producing relay entry for the given relay
+request. 
+
+There is a governable timeout for DKG to complete and for the result to be
+submitted. DKG timeout includes the time it takes to execute off-chain protocol
+to generate a key, and the time it takes for all group members to become
+eligible to submit the result and failing to do so. Note that unlike in the case
+of relay entry, `RandomBeacon.submitDkgResult(DKG.Result calldata dkgResult)`
+function enforces the eligibility of submitters on-chain. When DKG timeout is
+hit, anyone can call `RandomBeacon.notifyDkgTimeout()` function and receive the
+notifier's reward. The function unlocks the sortition pool and clears up DKG
+data but no slashing for DKG timeout is executed and no one is losing any
+rewards.
+
+[[fees-and-rewards]]
+=== Fees and Rewards
+
+Relay requester should provide a fee in T. The value of the fee is a governable
+parameter. The entire fee is deposited in the DKG rewards pool that is used to
+reimburse for different actions related to DKG.
+
+There is a fixed, governable reward for submitting and approving (after the
+challenge period!) DKG result paid from the DKG rewards pool. The reward is paid
+to the DKG result submitter in the transaction approving the DKG result. If the
+DKG result submitter failed to approve the result after the challenge period,
+anyone can do that and receive the submitter's reward.
```

Comment:
So the reward is paid to the DKG submitter or to the caller of the approve function?
But maybe describing it this way is more clear.

Reply:
The caller but DKG submitter is expected to be the caller of this function. Only the DKG result submitter can initially call it. If they do not, after a certain number of blocks, anyone can do that and claim the reward.